<a href="https://colab.research.google.com/github/RPGraciotti/BootCampAlura/blob/main/Modulo_6/Aula_1_Contextualiza%C3%A7%C3%A3o_do_mercado_financeiro.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Sessão 1 - Contexto financeiro
Introdução:

Pilares da cadeia de crédito:

1. Prospecção
2. Concessão - Foco do módulo do BootCamp
3. Gestão de Risco 
4. Recuperação

Pedido de crédito:
Instituição financeira utiliza um modelo de decisão + política de crédito

Bytebank é o banco fictício que a Alura utiliza para ferramentas didáticas e será trabalhado durante o curso.
Streamlit é uma biblioteca python que é utilizada para hospedar o modelo e gera uma interface sem que seja necessário conhecimento de front-end.

Projeto final: Simulador de avaliação de crédito utilizando o streamlit.

Objetivo é derivado de um desafio do kaggle, traduzido e manipulado para utilização: https://www.kaggle.com/rikdifos/credit-card-approval-prediction

# Sessão 2 - Primeiros passos com a base

In [1]:
import pandas as pd

In [2]:
df_cad = pd.read_csv("https://github.com/RPGraciotti/BootCampAlura/raw/main/Data/clientes_cadastrados.txt")
df_cad

ID_Cliente Genero  ...  Tamanho_Familia  Anos_empregado
0          5008804      H  ...              2.0            12.4
1          5008805      H  ...              2.0            12.4
2          5008806      H  ...              2.0             3.1
3          5008808      M  ...              1.0             8.4
4          5008809      M  ...              1.0             8.4
...            ...    ...  ...              ...             ...
438552     6840104      H  ...              1.0         -1000.7
438553     6840222      M  ...              1.0             8.2
438554     6841878      M  ...              1.0             1.0
438555     6842765      M  ...              2.0         -1000.7
438556     6842885      M  ...              2.0             3.3

[438557 rows x 18 columns]

In [3]:
df_cad.shape

(438557, 18)

In [4]:
df_cad.head()

ID_Cliente Genero  Tem_Carro  ...   Ocupacao  Tamanho_Familia  Anos_empregado
0     5008804      H          1  ...     Outros              2.0            12.4
1     5008805      H          1  ...     Outros              2.0            12.4
2     5008806      H          1  ...  Seguranca              2.0             3.1
3     5008808      M          0  ...      Venda              1.0             8.4
4     5008809      M          0  ...      Venda              1.0             8.4

[5 rows x 18 columns]

In [5]:
df_cad.loc[:,"ID_Cliente"].value_counts().unique() # significa que só tem no máximo id com 1 repetição

array([2, 1])

In [6]:
print(df_cad.duplicated().sum())
print(df_cad.loc[:,"ID_Cliente"].duplicated(keep = False).sum())

0
94


In [7]:
df_cad[df_cad.loc[:,"ID_Cliente"].duplicated(keep = False)] # keep = True mantém só o primeiro id repetido, 
# usando keep = True mantemos todos os valores repetidos

ID_Cliente Genero  ...  Tamanho_Familia  Anos_empregado
421211     7702516      M  ...              4.0             3.4
421268     7602432      H  ...              2.0             3.6
421349     7602432      M  ...              2.0         -1000.7
421464     7836971      H  ...              3.0            15.1
421698     7213374      H  ...              2.0             2.6
...            ...    ...  ...              ...             ...
433158     7282535      M  ...              2.0         -1000.7
433159     7742853      H  ...              1.0             4.6
433217     7135270      M  ...              2.0         -1000.7
433666     7091721      M  ...              2.0             6.2
433789     7618285      M  ...              1.0             2.8

[94 rows x 18 columns]

# Desafio 1: Explorar IDs repetidos (Atualização de cadastro? Erro?)

Exemplo do cliente 7602432: Tem duas entradas, a segunda entrada tem valores muito discrepantes, levando a crer ser um erro. Mas o mais interessante é a coluna "Anos_empregado" dando um valor de -1000. Isso certamente parece um erro de preenchimento. Atentar para verificar se isso se repete.

In [9]:
ids_repetidos = df_cad[df_cad.loc[:,"ID_Cliente"].duplicated(keep = False)].loc[:,"ID_Cliente"]

In [15]:
ids_repetidos.index

Int64Index([421211, 421268, 421349, 421464, 421698, 421726, 421907, 422068,
            422077, 422119, 422326, 422660, 422747, 423076, 423143, 423160,
            423327, 423416, 423592, 423702, 423794, 423852, 423874, 424151,
            424224, 424480, 424618, 424708, 424819, 425023, 425235, 425306,
            425486, 425539, 425624, 425724, 426018, 426022, 426198, 426325,
            426488, 426563, 426665, 426714, 426818, 426845, 427007, 427016,
            427114, 427164, 427167, 427287, 427778, 427883, 428048, 428166,
            428168, 428171, 428620, 428632, 428802, 428911, 428986, 429029,
            429053, 429078, 429104, 429121, 429228, 429337, 429834, 430187,
            430284, 430436, 430746, 430756, 431201, 431391, 431414, 431545,
            431911, 432088, 432134, 432271, 432355, 432379, 432426, 432562,
            432643, 433158, 433159, 433217, 433666, 433789],
           dtype='int64')

In [16]:
df_cad_no_rep = df_cad.drop(ids_repetidos.index)

In [24]:
df_cad_no_rep.loc[:,"ID_Cliente"].value_counts()

6293503    1
5166449    1
6210931    1
5174645    1
5170551    1
          ..
7410037    1
6357367    1
6375806    1
5683584    1
6291456    1
Name: ID_Cliente, Length: 438463, dtype: int64

In [26]:
df_cad_no_rep.isnull().sum()

ID_Cliente               0
Genero                   0
Tem_Carro                0
Tem_Casa_Propria         0
Qtd_Filhos               0
Rendimento_Anual         0
Categoria_de_renda       0
Grau_Escolaridade        0
Estado_Civil             0
Moradia                  0
Idade                    0
Tem_Celular              0
Tem_telefone_trabalho    0
Tem_telefone_fixo        0
Tem_email                0
Ocupacao                 0
Tamanho_Familia          0
Anos_empregado           0
dtype: int64

In [27]:
df_cad_no_rep.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 438463 entries, 0 to 438556
Data columns (total 18 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   ID_Cliente             438463 non-null  int64  
 1   Genero                 438463 non-null  object 
 2   Tem_Carro              438463 non-null  int64  
 3   Tem_Casa_Propria       438463 non-null  int64  
 4   Qtd_Filhos             438463 non-null  int64  
 5   Rendimento_Anual       438463 non-null  float64
 6   Categoria_de_renda     438463 non-null  object 
 7   Grau_Escolaridade      438463 non-null  object 
 8   Estado_Civil           438463 non-null  object 
 9   Moradia                438463 non-null  object 
 10  Idade                  438463 non-null  float64
 11  Tem_Celular            438463 non-null  int64  
 12  Tem_telefone_trabalho  438463 non-null  int64  
 13  Tem_telefone_fixo      438463 non-null  int64  
 14  Tem_email              438463 non-nu

# Desafio 2: Explorar as colunas (features), identificar contínuas, categóricas e binárias